
i.MXL MMC controller has a built-in card detection. This patch makes it
available for the one who doesn't want to sacrifice a GPIO to do that.

Detection relies on SD_DAT3 pin being pulled high when a card is
inserted. So chip's internal pull-down resistor is activated to have
this pin low by default (recommended by MMC specs in any case).

Signed-off-by: Julien Boibessot <julien.boibessot@armadeus.com>
Signed-off-by: Eric Jarrige <jorasse@armadeus.org>
 
--- linux-org/drivers/mmc/host/imxmmc.c
+++ linux-imx/drivers/mmc/host/imxmmc.c
@@ -901,6 +901,27 @@
 	.get_ro		= imxmci_get_ro,
 };
 
+/* Controller built-in card detection */
+static int imxmci_card_present(void)
+{
+	static int current_state = 1;
+	static int nb_trials = 3;
+	int new_state = (MMC_STATUS & STATUS_CARD_PRESENCE) ? 1 : 0;
+
+	/* Take state into account only if data are not being transmitted */
+	if (!(MMC_STATUS & STATUS_CARD_BUS_CLK_RUN)
+			&& (new_state != current_state)) {
+		if (0 == --nb_trials) {
+			current_state = new_state;
+			nb_trials = 3;
+		}
+	} else {
+		nb_trials = 3;
+	}
+
+	return current_state;
+}
+
 static void imxmci_check_status(unsigned long data)
 {
 	struct imxmci_host *host = (struct imxmci_host *)data;
@@ -1022,6 +1041,10 @@
 	ret = request_irq(host->irq, imxmci_irq, 0, DRIVER_NAME, host);
 	if (ret)
 		goto out;
+
+	/* If no way of detecting card is given, use built-in method */
+	if (!host->pdata->card_present)
+		host->pdata->card_present = imxmci_card_present;
 
 	if (host->pdata && host->pdata->card_present)
 		host->present = host->pdata->card_present(mmc_dev(mmc));
--- linux-org/arch/arm/mach-imx/include/mach/imx-regs.h
+++ linux-imx/arch/arm/mach-imx/include/mach/imx-regs.h
@@ -172,7 +172,7 @@
 #define PB9_AF_MS_PI1        ( GPIO_PORTB | GPIO_AF | 9 )
 #define PB10_PF_SD_DAT2      ( GPIO_PORTB | GPIO_PF | GPIO_PUEN  | 10 )
 #define PB10_AF_MS_SCLKI     ( GPIO_PORTB | GPIO_AF | 10 )
-#define PB11_PF_SD_DAT3      ( GPIO_PORTB | GPIO_PF | 11 )
+#define PB11_PF_SD_DAT3      ( GPIO_PORTB | GPIO_PF | GPIO_PUEN | 11 )
 #define PB11_AF_MS_SDIO      ( GPIO_PORTB | GPIO_AF | 11 )
 #define PB12_PF_SD_CLK       ( GPIO_PORTB | GPIO_PF | 12 )
 #define PB12_AF_MS_SCLK0     ( GPIO_PORTB | GPIO_AF | 12 )
